### PR TITLE
[12.0] [MIG] account_payment_credit_card

### PR DESCRIPTION
**Pull request by nikul-serpentcs**
Credit Card Payments
=================
This module provides a way to record Credit Card purchases and to pay them the same way regular purchases are recorded (via Vendor Bills) and paid (via Vendor Payments).

By default, if vendor bills are used to record credit card purchases, the liability (AP) from the supplier/vendor must be moved to the credit card company via a manual journal entry.

Because a journal entry contains less information than a vendor bill, there may be a loss of information that affects book keeping accuracy and the ability to properly reconcile the AP amount when the credit card statement arrives and needs to be settled.
This method also requires an additional step not needed when a vendor bill is paid via bank or cash.

This module automates the creation of the manual journal entry required to move the AP to the credit card company, retains all information about the purchase in an invoice document, and leverages the standard purchase and payment workflow already in place for bank and cash payments.

It supports as many Credit Cards as are needed, configured in the same way an additional payment method would be (i.e. via the creation of a new journal).

To settle a Credit Card statement, the regular workflow to record a vendor payment is used – allowing the removal of charges not included in the statement.
Non-purchase transactions like fees for annual membership, balance transfers, cash advances and foreign transactions; as well as charges for late payments and returned checks; can be entered either as Vendor Invoices or manual Journal Entries as users elect.

Both of these methods will allow these items to be settled when making a payment to the credit card company.

This module also supports payment cancellation and re-entry (in the case a mistaken amount is entered) as well as refunds (where the purchase is returned and a credit from the credit card company will be issued).

### CODE REVIEW

**Comment by nikul-serpentcs** [comment]
@pedrobaeza  @max3903 @JayVora-SerpentCS @bodedra Could you please Review it?

File: account_payment_credit_card/readme/INSTALL.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/INSTALL.rst]: This file is not needed then

Comment (alexey-pelykh) [inline account_payment_credit_card/readme/INSTALL.rst]: This file is not needed in this case

File: account_payment_credit_card/views/account_view.xml
Comment (alexey-pelykh) [inline account_payment_credit_card/views/account_view.xml:3]: Not required, yet adding copyright & license to XML (which is kind-a code) would be great

File: account_payment_credit_card/models/account.py
Comment (alexey-pelykh) [inline account_payment_credit_card/models/account.py]: This is quite generic field name, maybe we should consider refactoring this?

File: account_payment_credit_card/models/account_move.py
Comment (alexey-pelykh) [inline account_payment_credit_card/models/account_move.py]: Just `super()` is ok

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: account_payment_credit_card/views/account_view.xml
Comment (bodedra) [reply account_payment_credit_card/views/account_view.xml:3]: @nikul-serpentcs Would you please copy copyright from [these file](https://github.com/OCA/purchase-workflow/blob/11.0/stock_landed_cost_company_percentage/views/res_company_view.xml#L3) and paste it here.

Comment (nikul-serpentcs) [reply account_payment_credit_card/views/account_view.xml:3]: @alexey-pelykh @bodedra  No need to here, In XML file not mandatory.

File: account_payment_credit_card/models/account_move.py
Comment (nikul-serpentcs) [reply account_payment_credit_card/models/account_move.py]: @alexey-pelykh yes, but in a generic format, we will keep this.

File: account_payment_credit_card/readme/INSTALL.rst
Comment (nikul-serpentcs) [reply account_payment_credit_card/readme/INSTALL.rst]: @alexey-pelykh If I use [this](https://github.com/OCA/maintainer-tools#readme-generator) then automatically generate this file.

> This file is not needed in this case

Yes right @alexey-pelykh, but It's not blocking any more, So keep here
In feture, any specific installation required then we will put here

**Review by AaronHForgeFlow — COMMENTED** [review]
Got this when validating a vendor bill on runbot:
image

**Comment by nikul-serpentcs** [comment]
@aheficent Done

**Comment by nikul-serpentcs** [comment]
@alexey-pelykh @aheficent Review Please.

File: account_payment_credit_card/readme/INSTALL.rst
Comment (alexey-pelykh) [reply account_payment_credit_card/readme/INSTALL.rst]: In that case, whoever would add the feature that requires special installation will add this file - right now this file is not needed.

File: account_payment_credit_card/models/account.py
Comment (alexey-pelykh) [reply account_payment_credit_card/models/account.py]: @nikul-serpentcs any comments on `partner_id`? Maybe something like `card_processor_partner_id`?

**Comment by alexey-pelykh** [comment]
@nikul-serpentcs added few more comments

File: account_payment_credit_card/models/account.py
Comment (nikul-serpentcs) [reply account_payment_credit_card/models/account.py]: @alexey-pelykh Done !!

**Review by bjeficent — COMMENTED** [review]
I have been testing the module and I'm not sure how to set the bill to paid. When i pay the vendor bill it is not reconciled with the payment. This should be detailed in the Usage section of the Readme.

**Comment by AaronHForgeFlow** [comment]
Same as @bjeficent Can't see how to pay the bill. Please clarify it in the usage section.

File: account_payment_credit_card/models/account.py
Comment (alexey-pelykh) [inline account_payment_credit_card/models/account.py:11]: @nikul-serpentcs please specify `oldname` during renaming for proper migration

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: account_payment_credit_card/models/account.py
Comment (nikul-serpentcs) [reply account_payment_credit_card/models/account.py:11]: @alexey-pelykh Done with `oldname`.

**Comment by nikul-serpentcs** [comment]
@aheficent @bjeficent Added Usage section with clarifying information.

If not understand then @max3903 Could you please describe more here.

**Comment by bodedra** [comment]
@bjeficent @aheficent For proper usage of these module, we have do some configuration. You may review it under configuration section: https://github.com/OCA/account-payment/pull/211/files#diff-0c86289c17ad2f849f2e41c684918cdfR53 Afterwards it will work smoothly. Reference:

screenshot from 2018-12-20 13-35-11

File: account_payment_credit_card/readme/USAGE.rst
Comment (bodedra) [inline account_payment_credit_card/readme/USAGE.rst:6]: IMHO Please remove repeated description section. All necessary information are given in configuration section. Reference: https://github.com/OCA/account-payment/pull/211/files#diff-0c86289c17ad2f849f2e41c684918cdfR49 

Agree ?

File: account_payment_credit_card/README.rst
Comment (bodedra) [inline account_payment_credit_card/README.rst:94]: IMO Two emails are not required to add.  :thought_balloon:

**Review by bodedra — CHANGES_REQUESTED** [review]

File: account_payment_credit_card/README.rst
Comment (nikul-serpentcs) [reply account_payment_credit_card/README.rst:94]: Why?

File: account_payment_credit_card/readme/USAGE.rst
Comment (nikul-serpentcs) [reply account_payment_credit_card/readme/USAGE.rst:6]: @bodedra No, `configuration section`  in only those information which is used for module configuration.
so need to remove this.

And `description section` in only describe module functionality.

**Comment by AaronHForgeFlow** [comment]
@bodedra I'll recheck again but I filled those fields in my previous attempt. Thanks for the notice.

File: account_payment_credit_card/readme/USAGE.rst
Comment (bodedra) [reply account_payment_credit_card/readme/USAGE.rst:6]: @max3903 Would you please check these ?

File: account_payment_credit_card/README.rst
Comment (bodedra) [reply account_payment_credit_card/README.rst:94]: @OCA/accounting-maintainers What's your call ?

Comment (AaronHForgeFlow) [reply account_payment_credit_card/README.rst:94]: I don't understand, what are two?

Comment (bodedra) [reply account_payment_credit_card/README.rst:94]: @aheficent Here is reference. 

screenshot from 2018-12-20 16-02-35

Comment (AaronHForgeFlow) [reply account_payment_credit_card/README.rst:94]: I see. I've seen that in other modules but I am not sure about this, let's see what other maintainers say.

Comment (JayVora-SerpentCS) [reply account_payment_credit_card/README.rst:94]: Contributor can be individuals, so IMHO its fine to add two persons from same firm. 3 names from same firm appear in the same file FYI.
This isn't a big issue though.

Comment (bodedra) [reply account_payment_credit_card/README.rst:94]: @JayVora-SerpentCS I don't meant with it. We can add as many persons from same firm. 3 names that you mentioned, each and every person contribute or improve something on these module. You may go through history. :wink: What I understand, If any new brand module introduce first time, can add many persons from same firm but not upgrade level. One contributor should add their own name on contributor list.

I haven't seen these before so comment it.

I will not stretch it any more :laughing: on these topic as not a big issue.

Let's move forward with @OCA/accounting-maintainers opinion

File: account_payment_credit_card/views/account_view.xml
Comment (bodedra) [reply account_payment_credit_card/views/account_view.xml:3]: @nikul-serpentcs yes. it's not mandatory. But I have improve it at https://github.com/nikul-serpentcs/account-payment/pull/1

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: It works fine on runbot. Pending to squash the commits a bit and attend other reviews though.

**Comment by JayVora-SerpentCS** [comment]
@nikul-serpentcs please squash commits.

**Comment by nikul-serpentcs** [comment]
> @nikul-serpentcs please squash commits.

@JayVora-SerpentCS It's Already squashed

**Comment by nikul-serpentcs** [comment]
@alexey-pelykh @pedrobaeza Could you please review it.

File: account_payment_credit_card/readme/CONFIGURE.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/CONFIGURE.rst]: Please remove empty line

File: account_payment_credit_card/models/account_move.py
Comment (alexey-pelykh) [inline account_payment_credit_card/models/account_move.py:3]: Please remove empty line, between header and first import only 1 empty line AFAIR

File: account_payment_credit_card/readme/CONTRIBUTORS.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/CONTRIBUTORS.rst]: Please remove empty line

File: account_payment_credit_card/readme/CREDITS.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/CREDITS.rst]: Please remove empty line

File: account_payment_credit_card/readme/DESCRIPTION.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/DESCRIPTION.rst]: Please remove empty line

File: account_payment_credit_card/readme/USAGE.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/USAGE.rst]: Please remove empty line

File: account_payment_credit_card/readme/USAGE.rst
Comment (alexey-pelykh) [inline account_payment_credit_card/readme/USAGE.rst]: Please remove empty line

**Review by alexey-pelykh — CHANGES_REQUESTED** [review]

File: account_payment_credit_card/README.rst
Comment (max3903) [reply account_payment_credit_card/README.rst:94]: Contributors for individuals. Credits for organizations.
https://github.com/OCA/maintainer-tools/tree/master/template/module/readme

File: account_payment_credit_card/readme/USAGE.rst
Comment (max3903) [reply account_payment_credit_card/readme/USAGE.rst:6]: First, let's split the README.rst file in the readme directory...

Comment (alexey-pelykh) [reply account_payment_credit_card/readme/USAGE.rst:6]: It's already separated, seems so

File: account_payment_credit_card/README.rst
Comment (pedrobaeza) [reply account_payment_credit_card/README.rst:94]: But credits are for extra referrals, like funders, not for the company.

**Comment by nikul-serpentcs** [comment]
@alexey-pelykh Done!!

File: account_payment_credit_card/readme/USAGE.rst
Comment (alexey-pelykh) [reply account_payment_credit_card/readme/USAGE.rst]: @nikul-serpentcs this one is still there

Comment (nikul-serpentcs) [reply account_payment_credit_card/readme/USAGE.rst]: > @nikul-serpentcs this one is still there

It's the Last line for this file, that's why not removed
If I removed then throw Error `add a new blank line`

Comment (alexey-pelykh) [reply account_payment_credit_card/readme/USAGE.rst]: Odd, this one doesn't have an empty line in the end - https://github.com/OCA/maintainer-tools/blame/master/template/module/readme/USAGE.rst - only newline. If possible, please provide more details on the error - which tool gives it

Comment (nikul-serpentcs) [reply account_payment_credit_card/readme/USAGE.rst]: > Odd, this one doesn't have an empty line in the end - https://github.com/OCA/maintainer-tools/blame/master/template/module/readme/USAGE.rst - only newline. If possible, please provide more details on the error - which tool gives it

In .py and .xml file get Error, So I added a blank line for that
anyway, I removed It, review it!

**Review by alexey-pelykh — APPROVED** [review]

**Comment by max3903** [comment]
Superseeded by #218

**Pull request closed without merging**
